### DISCUSSION TITLE
Time allowance on github & workflows

### DISCUSSION THREAD

**Discussion by iincer** [pacti-org/pacti#270] [General]
#### Multipliers 

We have to be careful with time allowances. The latest run of the production workflow [took about 40 minutes](https://github.com/FormalSystems/pacti/actions/runs/4445987635/usage?pr=267). The development run [took 3 minutes](https://github.com/FormalSystems/pacti/actions/runs/4445941619/usage?pr=267). However, these numbers do not correspond to GitHub's billing mechanism. Running jobs on Windows and macOS incurs in a x2 and x10 minute multiplier. This is our current usage for this month:

![image](https://user-images.githubusercontent.com/42584466/225945582-d8fc97e8-6ead-4c7a-ac9f-a4f170f2884d.png)

To account for the 320 used minutes, we have to multiply the macOS minutes by 10 and the Windows minutes by 2.

Our latest production run took 12, 18, and 13 minutes on ubuntu, Windows, and macOS, respectively, yielding a total of 178 weighted minutes. We would be able to run this workflow a few times per month, as the free version of GitHub gives us 2000 minutes per month for private repos.

#### Public repos

[Here is GitHub's pricing model](https://docs.github.com/en/billing/managing-billing-for-github-actions/about-billing-for-github-actions). This is what it says regarding public repos:

> GitHub Actions usage is free for standard GitHub-hosted runners in public repositories, and for self-hosted runners. 
> Note: Entitlement minutes cannot be used for Windows and Ubuntu runners over 2-cores. These runners will always be charged for, including in public repos. For more information, see "[About billing for GitHub Actions](https://docs.github.com/en/billing/managing-billing-for-github-actions/about-billing-for-github-actions#per-minute-rates)."

_Originally posted by @iincer in https://github.com/FormalSystems/pacti/issues/267#issuecomment-1474041900_

**Comment by ayush9pandey** [comment]
We have continuous integration in Pacti now, where some workflows are automatic and some others are manual. Instructions for all developers:

1. If you are in the process of developing a feature/fixing a bug in your own development branch: Make sure that the `Pacti Development` workflow passes. This will automatically be run with every commit, so you should be getting notifications of what Pacti functionality failed (if any). 
2. If you are contributing a development feature to any branch (that is, your development has finished): Run the `Pacti Development Review` workflow. This will test the functionality and also code quality.
3. If you are reviewing a pull-request to the main branch from another developer: Make sure to go to the "Actions" tab and run the `Pacti Pre-Production` workflow manually. This workflow tests the package for functionality and quality on two Python versions. 

Until the repo is private, we may choose to run the `Pacti Production` workflow once a month to make sure that the software works with all OS and all Python versions.

**Comment by ayush9pandey** [comment]
For public repositories, I see the same 2000 allowed minutes in the billing section. Not sure what's the difference then between public and private..

**Comment by iincer** [comment]
From what they say about their pricing model, I think the difference is that Ubuntu and Windows running on at most two cores are free if the repo is public. For anything else they will bill us.

**Reply by ayush9pandey** [reply]
That makes sense, thanks. It's not clear to me how the number of cores are decided. Is it possible to run on a single core for MacOS for example?
